packages/solo/src/chain-cosmos-sdk.js

```diff
@@ -381,137 +330,210 @@ ${chainID} chain does not yet know of address ${myAddr}${adviseEgress(myAddr)}
   // Begin the block notifier cycle.
   const blockNotifier = getBlockNotifier();
 
-  /**
-   * This function is entered at most the same number of times
-   * as the blockNotifier announces a new block.
-   *
-   * It then gets the mailbox.  There are no optimisations.
-   *
-   * @param {number=} lastBlockUpdate
-   */
-  const recurseEachNewBlock = (lastBlockUpdate = undefined) => {
-    blockNotifier
-      .getUpdateSince(lastBlockUpdate)
-      .then(({ updateCount, value }) => {
-        assert(value, X`${GCI} unexpectedly finished!`);
-        log.debug(`new block on ${GCI}, fetching mailbox`);
-        getMailbox()
-          .then(ret => {
-            if (!ret) {
-              // The getMailbox was cancelled.
-              return;
-            }
-
-            const { outbox, ack } = ret;
-            // console.debug('have outbox', outbox, ack);
-            inbound(GCI, outbox, ack);
-          })
-          .catch(e => log.error(`Failed to fetch ${GCI} mailbox:`, e));
-        recurseEachNewBlock(updateCount);
-      });
-  };
-
-  // Begin the block consumer.
-  recurseEachNewBlock();
+  const { notifier: sendNotifier, updater: sendUpdater } = makeNotifierKit();
 
   let totalDeliveries = 0;
-  async function deliver(newMessages, acknum) {
+  let messagePool = [];
+  let highestAck = -1;
+  let sequenceNumber = 0n;
+  const sendFromMessagePool = async () => {
     let tmpInfo;
+    const messages = messagePool;
+    messagePool = [];
     try {
       totalDeliveries += 1;
-      log(
+      console.log(
         `delivering to chain (trips=${totalDeliveries})`,
         GCI,
-        newMessages,
-        acknum,
+        messages,
+        highestAck,
       );
 
-      // Peer and submitter are combined in the message format (i.e. we removed
-      // the extra 'myAddr' after 'tx swingset deliver'). All messages from
-      // solo vats are "from" the signer, and messages relayed from another
-      // chain will have other data to demonstrate which chain it comes from
+      // TODO: remove this JSON.stringify: change 'deliverMailboxReq' to have
+      // more structure than a single string, and have the CLI handle variable
+      // args better
+      tmpInfo = await makeTempFile(
+        'ag-solo-cosmos-deliver.',
+        JSON.stringify([messages, highestAck]),
+      );
 
-      // TODO: remove this JSON.stringify([newMessages, acknum]): change
-      // 'deliverMailboxReq' to have more structure than a single string, and
-      // have the CLI handle variable args better
+      // Deliver message over file, as it could be big.
+      let args = ['tx', 'swingset', 'deliver', `@${tmpInfo.path}`];
+      if (clientAddr !== helperAddr) {
```

Comment:
this might be easier to read if the `--generate-only` call were in a subroutine, and there were two branches here, that ended with the `args=` assignment

Reply:
I cleaned this up with a helper function but kept a single branch.  It should be better now.